### ISSUE TITLE
Access Token refresh can be very slow

### ISSUE THREAD

**Issue by fowlerp-qlik** [AzureAD/azure-activedirectory-library-for-android#1415]
[fromAdal.log](https://github.com/AzureAD/azure-activedirectory-library-for-android/files/3004571/fromAdal.log)

It is expected that after one hour of inactivity getting a new access token will take less than 3 seconds. At a large customer we this time can be as high as 13 seconds!  

I cannot reproduce this is our own Azure tenant though. 

The attached ADAL trace logs show that a "discovery" request is performed to https://login.microsoft.com/common/discovery/instance.

On average these requests take 300-400 milliseconds. There is no parameter that I can set to avoid this request as the tester (by design) is killing our App then relaunching to obtain these measurements.

At 7:35am the request took roughly 3.7 seconds.

Also the attached logs show an occasional delay after connecting the broker service. At 7:35am this delay was roughly 10 seconds! But I see other such delays of 2.5, 2.4 and 3.8 seconds in the logs.

The device under test is a Redmi Note 5 Pro running Android 8.1. It is connected to the internet over LTE. The phone is in India as is their connector and application server. Customer states that their Android Xamarin based LOB apps do not experience this delay.

Our code is setting up an AuthenticationContext with an authority of 
https://login.microsoftonline.com/common

Customer has a Conditional Access rule such that the webapi we are obtaining a token for (the Audience) is only accessible from compliant devices.

**Comment by shoatman** [comment]
Hi @fowlerp-qlik - I see the delay that you're seeing in the logs.  Outlook for Android has the ability to collect diagnostic logs and transmit them to us.  It's within Settings->Help & Feedback -> Collect Diagnostics.  If you send us the related incident number we'll be able to see which broker host app is acting as broker (Company Portal or Micorsoft Authenticator) and be able to see the logs from those apps... to see why this request is taking so long.

**Comment by fowlerp-qlik** [comment]
The broker is MS Authenticator:

AcquireTokenWithBrokerRequest:logBrokerVersion:2019-03-25 05:46:40-e83a9c35-3932-444e-b520-2651634f3c6e-Broker app is: com.azure.authenticator;Broker app version: VersionName=6.4.5;VersonCode=148. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-854075024]<=

Do you mean that the Outlook diagnostics will include an incident number and further the logs will include logs from that broker?

**Comment by shoatman** [comment]
That's correct.  Outlook's collect diagnostics will collect logs from the broker host application.  Which will include the broker logs... such that we can see what's slow inside the broker.

**Comment by fowlerp-qlik** [comment]
ok. Customer has reproduced the issue and reported the following:

"I did two rounds of testing. This time it was not taking 20+ seconds. But still it took 14s at 11:06pm  and 16s at 12:10am."

The incident ID is 8D32BY86.

**Comment by shoatman** [comment]
Thank you.  Appreciated we'll take a look.

**Comment by fowlerp-qlik** [comment]
Hi Shane, What are your initial findings in the logs?

**Comment by shoatman** [comment]
Did you upload the logs from Authenticator or from Outlook?  It looks like I only have the authenticator logs and according to those logs the Authenticator is not the active broker on the device where tested.  I don't see broker log output in the authenticator logs.

**Comment by fowlerp-qlik** [comment]
The originally attached logs (fromAdal.log) are the ADAL library trace logs from our Android App. 
I have not attached any others. I had thought that the Outlook collect logs diagnostic would push logs to a server and the incident Id would be the pointer to those logs. 

So you are saying the customer needs to more explicitly send the logs (to me)?

**Comment by fowlerp-qlik** [comment]
ping ..

**Comment by shoatman** [comment]
From what I can see the customer used the Microsoft Authenticator app to initiate the upload of the logs.  Instead of Outlook for Android.  Please ask them to use Outlook for Android and have them provide the new incident id.  The existing incident id only contained the Microsoft Authenticator logs and there was nothing in those logs to suggest that the authentication broker was being invoked.  Outlook for Android will collect logs for Company Portal , Microsoft Authenticator, etc...

**Comment by fowlerp-qlik** [comment]
Customer created a new ticket from Outlook log. The incident id: B8AC5DBE
 
It is not clear what the ADAL token refresh duration was though .. hoping it was long enough to show an issue in a specific step of the process.

**Comment by fowlerp-qlik** [comment]
Customer tested again and created incident FAC6A9A1. Logs send via MS Outlook for Android. The App's ADAL trace logs show a network delay (customer says network was "bad" at the time). But logs also show a delay starting at 14:10:13,158 for a duration of some 12 seconds.

localhost_QSM.log:2019-04-02 14:09:30,641-INFO From ADAL: AuthenticationContext:getRedirectUriForBroker:2019-04-02 08:39:30-null-Get expected redirect Uri.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-945817232]<=
localhost_QSM.log:2019-04-02 14:09:30,641-INFO From ADAL: BrokerProxy:2019-04-02 08:39:30-null-Device runs on 23 and above, skip the check for 22 and below. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-945817232]<=
localhost_QSM.log:2019-04-02 14:09:30,777-INFO From ADAL: AcquireTokenRequest:acquireToken:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Sending async task from thread:2496 ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-945817232]<=
localhost_QSM.log:2019-04-02 14:09:30,778-INFO From ADAL: AcquireTokenRequest:acquireToken:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Running task in thread:2650 ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,782-INFO From ADAL: Discovery:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Sending discovery request to query url.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,783-INFO From ADAL: WebRequestHandler:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-WebRequestHandler thread2650 ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,783-INFO From ADAL: HttpWebRequest:send:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-HttpWebRequest send.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,784-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-HttpWebRequest setupConnection. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,794-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Setting header.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,794-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Setting header.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,795-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Setting header.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,795-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Setting header.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,795-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Setting header.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:09:30,796-INFO From ADAL: HttpWebRequest:setupConnection:2019-04-02 08:39:30-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Setting header.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,727-INFO From ADAL: HttpWebRequest:send:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Response is received. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,737-INFO From ADAL: AuthenticationContext:getRedirectUriForBroker:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Get expected redirect Uri.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,740-INFO From ADAL: AcquireTokenRequest:verifyBrokerRedirectUri:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-The broker redirect URI is valid. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,741-INFO From ADAL: AcquireTokenRequest:tryAcquireTokenSilent:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Try to acquire token silently, return valid AT or use RT in the cache. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,742-INFO From ADAL: AcquireTokenRequest:tryAcquireTokenSilentLocally:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Try to silently get token from local cache. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,745-INFO From ADAL: TokenCacheAccessor:getATFromCache:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-No access token exists. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,745-INFO From ADAL: AcquireTokenSilentHandler:getAccessToken:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-No valid access token exists, try with refresh token. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,747-INFO From ADAL: AcquireTokenSilentHandler:tryRT:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Regular token cache entry does not exist, try with MRRT. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,749-INFO From ADAL: AcquireTokenSilentHandler:tryMRRT:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-MRRT token does not exist, try with FRT ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,752-INFO From ADAL: AcquireTokenSilentHandler:tryFRT:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-FRT cache item does not exist, fall back to try MRRT. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,753-INFO From ADAL: AcquireTokenSilentHandler:useMRRT:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Send request to use MRRT for new AT. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,754-INFO From ADAL: AcquireTokenSilentHandler:useMRRT:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-MRRT does not exist, cannot proceed with MRRT for new AT. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,763-INFO From ADAL: AcquireTokenRequest:removeTokensForUser:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-No token items need to be deleted for the user. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,767-INFO From ADAL: AcquireTokenWithBrokerRequest:logBrokerVersion:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Broker app is: com.azure.authenticator;Broker app version: VersionName=6.4.7;VersonCode=150. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,768-INFO From ADAL: AcquireTokenWithBrokerRequest:acquireTokenWithBrokerSilent:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-User is specified for background(silent) token request, trying to acquire token silently. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,771-INFO From ADAL: BrokerAccountServiceHandler:bindToBrokerAccountService:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Binding to BrokerAccountService for caller uid.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:12,795-INFO From ADAL: BrokerAccountServiceHandler:bindToBrokerAccountService:2019-04-02 08:40:12-dc87eac7-1bc9-4f73-b600-c9d97177de5f-The status for brokerAccountService bindService call is: true ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=
localhost_QSM.log:2019-04-02 14:10:13,158-INFO From ADAL: BrokerAccountServiceHandler:2019-04-02 08:40:13-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Broker Account service is connected. ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-363101020]<=
localhost_QSM.log:2019-04-02 14:10:25,050-INFO From ADAL: AcquireTokenRequest:tryAcquireTokenSilent:2019-04-02 08:40:25-dc87eac7-1bc9-4f73-b600-c9d97177de5f-Token is successfully returned from silent flow.  ver:1.14.1-null =>void Java_com_qlik_qliksense_mobile_OAuthManagerImpl_logToQt(JNIEnv *, jobject, jstring)(60) [-1002972816]<=

**Comment by fowlerp-qlik** [comment]
Customer created one more ticket (incident id: 75D6CD0C). Time (2019-04-02 18:50:54,847)

**Comment by shoatman** [comment]
Thanks @fowlerp-qlik.  This definitely looks like a network reliability issue.  I see a number of failed requests due to the network connection being observed as unavailable as well as messages indicating that login.microsoftonline.com was not able to be resolved as an Address.  These messages appear during attempts to refresh a token via acquireTokenSilent.

When I pull the logs for the eventual successful request.  The latency within our servers is quite low (approx 300 ms).  There is retry logic in place within both ADAL and the Broker.  It's possible that the cumulative effect of these retries and slow/intermittent network connections is resulting in what you're observing.  I'll talk to some members of my team to see what if anything that we can propose to further diagnose what you are observing.

**Comment by fowlerp-qlik** [comment]
ok .. could you look at the 75D6CD0C as well. They say that the ADAL slowness occurs on reliable networks as well. Slowness is only on Android and only occurs after one hour of inactivity .. ie. the access token must be refreshed.

**Comment by fowlerp-qlik** [comment]
Customer asks that you look at the times around given the first incident ID corresponds to a time of network instability.

2019-04-02 16:30:00,238
2019-04-02 18:50:54,847

**Comment by shoatman** [comment]
I will, but likely not until later today tomorrow.  Dedicating my first 45 minutes to look at this github issues, but unfortunately have other things that I also need to look at.  Thanks for your understanding and patience.

**Comment by shoatman** [comment]
@sdabbiru - This is the issue that I was referring to.  

@fowlerp-qlik - I don't see anything unusual in the logs at that time.  I do see network failures earlier in the day.   I'm going to inquire about getting a build of Authenticator with verbose logging enabled.  To help us get more information.

**Comment by shoatman** [comment]
@danieldobalian - this is the issue that i'd like a special build of authenticator to help investigate.  What's the right way to make this available to @fowlerp-qlik and his customer?

**Comment by shoatman** [comment]
@fowlerp-qlik - Please email me directly and we can arrange to arrange providing you with a build with verbose logging enabled.

**Comment by iambmelt** [comment]
Followed up offline. Closing as resolved
